cli/internal/scope/scope_test.go

```diff
@@ -0,0 +1,207 @@
+			[]string{"@sample/*"},
+			util.Set{"@sample/app": "@sample/app"},
+		},
+		{
+			"return an array of matches",
+			[]string{"foo", "bar", "baz"},
+			[]string{"f*"},
+			util.Set{"foo": "foo"},
+		},
+		{
+			"return an array of matches",
+			[]string{"foo", "bar", "baz"},
+			[]string{"f*", "bar"},
+			util.Set{"bar": "bar", "foo": "foo"},
+		},
+		{
+			"return matches in the order the list were defined",
+			[]string{"foo", "bar", "baz"},
+			[]string{"*a*", "!f*"},
+			util.Set{"bar": "bar", "baz": "baz"},
+		},
+	}
+
+	for i, tc := range cases {
+		t.Run(fmt.Sprintf("%d-%s", i, tc.Name), func(t *testing.T) {
+			actual, err := getScopedPackages(tc.PackageNames, tc.Pattern)
+			if err != nil {
+				t.Fatalf("invalid scope parse: %#v", err)
+			}
+			assert.EqualValues(t, tc.Expected, actual)
+		})
+	}
+
+	t.Run(fmt.Sprintf("%d-%s", len(cases), "throws an error if no package matches the provided scope pattern"), func(t *testing.T) {
+		_, err := getScopedPackages([]string{"foo", "bar"}, []string{"baz"})
+		assert.Error(t, err)
+	})
+}
+
+type mockSCM struct {
+	changed []string
+}
+
+func (m *mockSCM) ChangedFiles(fromCommit string, includeUntracked bool, relativeTo string) []string {
+	changed := []string{}
+	for _, change := range m.changed {
+		if strings.HasPrefix(change, relativeTo) {
+			changed = append(changed, change)
+		}
+	}
+	return changed
+}
+
+func TestResolvePackages(t *testing.T) {
+	tui := ui.Default()
+	logger := hclog.Default()
+	// app1 -> libA -> libB
+	//
+	//        / libB
+	// app2 <
+	//        \ libC
+	//
+	graph := dag.AcyclicGraph{}
+	graph.Add("app1")
+	graph.Add("app2")
+	graph.Add("libA")
+	graph.Add("libB")
+	graph.Add("libC")
+	graph.Connect(dag.BasicEdge("libA", "libB"))
+	graph.Connect(dag.BasicEdge("app1", "libA"))
+	graph.Connect(dag.BasicEdge("app2", "libB"))
+	graph.Connect(dag.BasicEdge("app2", "libC"))
+	scc := dag.StronglyConnected(&graph.Graph)
+	packagesInfos := map[interface{}]*fs.PackageJSON{
+		"app1": {
+			Dir: "app/app1",
+		},
+		"app2": {
+			Dir: "app/app2",
+		},
+		"libA": {
+			Dir: "libs/libA",
+		},
+		"libB": {
+			Dir: "libs/libB",
+		},
+		"libC": {
+			Dir: "libs/libC",
+		},
+	}
+	packageNames := []string{}
+	for name := range packagesInfos {
+		packageNames = append(packageNames, name.(string))
+	}
+
+	testCases := []struct {
+		changed             []string
+		expected            []string
+		scope               []string
+		since               string
+		ignore              string
+		globalDeps          []string
+		includeDependencies bool
+		includeDependents   bool
+	}{
+		{
+			changed:  []string{"libs/libB/src/index.ts"},
+			expected: []string{"libB"},
+			since:    "dummy",
+		},
+		{
+			changed:  []string{"libs/libB/src/index.ts"},
+			expected: []string{},
+			since:    "dummy",
+			ignore:   "libs/libB/**/*.ts",
+		},
+		{
+			// a non-dependent lib changed
+			changed:  []string{"libs/libC/src/index.ts"},
+			expected: []string{},
```

Comment:
I think this is correct, given that we've supplied a scope which does not depend on the changed package. However our current scope resolution logic includes all changed packages in this situation.

Reply:
agreed